File: solutions/cpp/robot-simulator/2/robot_simulator.cpp
#include "robot_simulator.h"
#include <string>
#include <tuple>

namespace robot_simulator {

      
    Robot::Robot(position_t initial_position, Bearing initial_bearing) : my_position(initial_position), my_bearing(initial_bearing) {} 

        
    void Robot::execute_sequence(const std::string& sequence){
        for (auto it = sequence.begin(); it != sequence.end(); it++) {
            switch(*it) {
                case 'A':
                    advance(); 
                    break;
                case 'R':
                    turn_right();
                    break;
                case 'L':
                    turn_left();
                    break;
                default:
                    break;
            }    
        }
    }

    void Robot::turn_left(){
        switch(my_bearing) {
            case Bearing::NORTH:
                my_bearing = Bearing::WEST;
                break;
            case Bearing::WEST:
                my_bearing = Bearing::SOUTH;
                break;
            case Bearing::SOUTH:
                my_bearing = Bearing::EAST;
                break;
            default:
                my_bearing = Bearing::NORTH;
        }
    }

    void Robot::turn_right(){
        switch(my_bearing) {
            case Bearing::NORTH:
                my_bearing = Bearing::EAST;
                break;
            case Bearing::EAST:
                my_bearing = Bearing::SOUTH;
                break;
            case Bearing::SOUTH:
                my_bearing = Bearing::WEST;
                break;
            default:
                my_bearing = Bearing::NORTH;
        }
    }

    void Robot::advance(){
        switch(my_bearing) {
            case Bearing::NORTH:
                my_position.second += 1;
                break;
            case Bearing::EAST:
                my_position.first +=1;
                break;
            case Bearing::SOUTH:
                my_position.second -=1;
                break;
            default:
                my_position.first -=1;
        }
    }


}  // namespace robot_simulator

File: solutions/julia/robot-simulator/2/robot-simulator.jl
import Base: convert

@enum Heading NORTH=1 EAST=2 SOUTH=3 WEST=4 

struct Point
    x::Int64
    y::Int64
end

convert(Point,xy::Tuple{Int64, Int64})=Point(xy[1],xy[2])

mutable struct Robot
    position::Point 
    heading::Heading
end

position(r::Robot) = r.position
heading(r::Robot) = r.heading

function turn_left!(robot::Robot)
    rotated = circshift(collect(instances(Heading)), 1)
    robot.heading = rotated[UInt(robot.heading)]
    robot
end

function turn_right!(robot::Robot)    
    rotated = circshift(collect(instances(Heading)), -1)
    robot.heading = rotated[UInt(robot.heading)]
    robot
end

function advance!(robot::Robot)
    dx, dy = 0,0
    robot.heading==NORTH && (dy += 1)
    robot.heading==SOUTH && (dy -= 1)
    robot.heading==EAST && (dx += 1)
    robot.heading==WEST && (dx -= 1)

    robot.position = Point(robot.position.x + dx,
                          robot.position.y + dy)
    robot
end

function move!(robot::Robot, instructions)
    for instruction in instructions
        if instruction=='L'
            turn_left!(robot)
        elseif instruction=='R'
            turn_right!(robot)
        else #instruction=='A'
            advance!(robot)
        end
    end
    robot
end